lib/orangelight/browse_lists/call_number_csv.rb

```diff
@@ -0,0 +1,131 @@
+            body = solr_cursor_page_body(cursor_mark)
+            body['response']['docs'].each do |record|
+              next unless record[facet_field]
+              record[facet_field].each do |cn|
+                sort_cn = StringFunctions.cn_normalize(cn)
+                next if multi_cn_lookup.key?(sort_cn)
+                csv << parse_cn_row(record, cn, sort_cn)
+              end
+            end
+            next_cursor_mark = body['nextCursorMark']
+            break if cursor_mark == next_cursor_mark
+            cursor_mark = next_cursor_mark
+          end
+        end
+      end
+
+      # Get a page of solr results for actual items starting at the given number
+      def solr_cursor_page_body(cursor_mark)
+        retries = 0
+        req = {}
+        cn_fields = "#{facet_field},title_display,title_vern_display,author_display,author_s,id,pub_created_vern_display,pub_created_display,holdings_1display"
+        loop do
+          cn_request = "#{core_url}select?q=*%3A*&fl=#{cn_fields}&wt=json&indent=true&defType=edismax&facet=false&sort=id%20asc&rows=#{rows}&cursorMark=#{cursor_mark}"
+          resp = conn.get cn_request.to_s
+          req = JSON.parse(resp.body)
+          break if req['response']
+          Rails.logger.error "Call number browse generation failed at iteration with cursor mark #{cursor_mark}. Response from solr was: #{resp}"
+          raise SolrResponseError if retries >= 2
+          retries += 1
+        end
+        req
+      end
+
+      def parse_cn_row(record, cn, sort_cn)
+        bibid = record['id']
+        title = record['title_display']
+        if record['title_vern_display']
+          title = record['title_vern_display']
+          dir = title.dir
+        else
+          dir = 'ltr' # ltr for non alt script
+        end
+        if record['pub_created_vern_display']
+          date = record['pub_created_vern_display'][0]
+        elsif record['pub_created_display'].present?
+          date = record['pub_created_display'][0]
+        end
+        label = cn
+        if record['author_display']
+          author = record['author_display'][0..1].last
+        elsif record['author_s']
+          author = record['author_s'][0]
+        end
+        if record['holdings_1display']
+          holding_block = JSON.parse(record['holdings_1display'])
+          holding_record = holding_block.select { |_k, h| h['call_number_browse'] == cn }
+          unless holding_record.empty?
+            if multiple_locations?(holding_record)
+              location = 'Multiple locations'
+            else
+              holding_id = holding_record.keys.first
+              location = holding_record[holding_id]['location']
+            end
+          end
+        end
+        holding_id ||= ''
+        location ||= ''
+        [sort_cn, label, dir, '', title, author, date, bibid, holding_id, location]
+      end
+
+      # determines if there are multiple locations for the same call number and same bib
```

Comment:
thought: If you extract a record object above then this method could be a private method on that object.